python/dgl/distributed/sparse_emb.py

```diff
@@ -58,148 +60,101 @@ class DistEmbedding:
 
     Note
     ----
-    When a ``DistEmbedding``  object is used when the deep learning framework is recording
-    the forward computation, users have to invoke py:meth:`~dgl.distributed.SparseAdagrad.step`
+    When a ``NodeEmbedding``  object is used when the deep learning framework is recording
+    the forward computation, users have to invoke py:meth:`~dgl.distributed.optim.SparseAdagrad.step`
     afterwards. Otherwise, there will be some memory leak.
     '''
     def __init__(self, num_embeddings, embedding_dim, name=None,
                  init_func=None, part_policy=None):
         self._tensor = DistTensor((num_embeddings, embedding_dim), F.float32, name,
-                                  init_func, part_policy)
+                                  init_func=init_func)
         self._trace = []
-
-    def __call__(self, idx):
+        self._name = name
+        self._num_embeddings = num_embeddings
+        self._embedding_dim = embedding_dim
+
+        # Check whether it is multi-gpu training or not.
+        if th.distributed.is_initialized():
+            self._rank = th.distributed.get_rank()
+            self._world_size = th.distributed.get_world_size()
+        else:
+            assert 'th.distributed shoud be initialized'
+        self._optm_state = None # track optimizer state
+        self._part_policy = part_policy
+
+    def __call__(self, idx, device=th.device('cpu')):
         idx = utils.toindex(idx).tousertensor()
-        emb = self._tensor[idx]
+        emb = self._tensor[idx].to(device, non_blocking=True)
         if F.is_recording():
             emb = F.attach_grad(emb)
-            self._trace.append((idx, emb))
+            self._trace.append((idx.to(device, non_blocking=True), emb))
         return emb
 
     def reset_trace(self):
         '''Reset the traced data.
         '''
         self._trace = []
 
-class SparseAdagradUDF:
-    ''' The UDF to update the embeddings with sparse Adagrad.
-
-    Parameters
-    ----------
-    lr : float
-        The learning rate.
-    '''
-    def __init__(self, lr):
-        self._lr = lr
+    @property
+    def part_policy(self):
+        return self._part_policy
 
-    def __call__(self, data_store, name, indices, data):
-        ''' Update the embeddings with sparse Adagrad.
+    @property
+    def rank(self):
+        """Return rank of current process.
 
-        This function runs on the KVStore server. It updates the gradients by scaling them
-        according to the state sum.
+        Returns
+        -------
+        int
+            The rank of current process.
+        """
+        return self._rank
```

Comment:
this rank is from pytorch distributed. it's confusing because DistGraph exposes the rank of DGL distributed.